backend/worker/error_scorer.go

```diff
@@ -0,0 +1,283 @@
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/highlight-run/highlight/backend/model"
+	"github.com/highlight-run/highlight/backend/redis"
+	"github.com/highlight-run/highlight/backend/store"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+type ErrorGroupWithLastObjectId struct {
+	ID                int
+	ProjectID         int
+	CreatedAt         time.Time
+	LastErrorObjectId int
+}
+
+type ErrorScorer struct {
+	store *store.Store
+	db    *gorm.DB
+	redis *redis.Client
+}
+
+func NewErrorScorer(store *store.Store, db *gorm.DB, redis *redis.Client) *ErrorScorer {
+	return &ErrorScorer{
+		store: store,
+		db:    db,
+		redis: redis,
+	}
+}
+
+func (errorScorer *ErrorScorer) ScoreImpactfulErrors(ctx context.Context) {
+	var errorGroups []*ErrorGroupWithLastObjectId
+
+	lastComputedErrorInstanceId, _ := errorScorer.redis.GetLastComputedImpactfulErrorObjectId(ctx)
+
+	if err := errorScorer.db.Raw(`
+		SELECT grp.id, grp.created_at, grp.project_id, MAX(obj.id) as last_error_object_id
+		FROM error_objects obj
+		INNER JOIN error_groups grp
+			ON grp.id = obj.error_group_id
+		WHERE obj.created_at >= now() - INTERVAL '10 minutes'
+			AND obj.id > ?
+		GROUP BY grp.id
+		ORDER BY last_error_object_id ASC
+	`, lastComputedErrorInstanceId).Scan(&errorGroups).Error; err != nil {
+		log.WithContext(ctx).Error(errors.Wrap(err, "error querying for error objects"))
+		return
+	}
+
+	if len(errorGroups) == 0 {
+		log.WithContext(ctx).WithField("lastErrorObjectId", lastComputedErrorInstanceId).Info("No errors to score")
+		return
+	}
+
+	// TODO(spenny): currently N+1 query
+	for _, errorGroup := range errorGroups {
+		occurranceScore, affectedUsersScore, err := errorScorer.scoreErrorGroup(ctx, errorGroup)
+		_ = errorScorer.redis.SetLastComputedImpactfulErrorObjectId(ctx, errorGroup.LastErrorObjectId)
+		if err != nil {
+			log.WithContext(ctx).WithField("errorGroupId", errorGroup.ID).Error(err)
+			continue
+		}
+
+		totalScore := occurranceScore + affectedUsersScore
+		currentTime := time.Now()
+
+		if err := errorScorer.db.Model(&model.ErrorGroup{Model: model.Model{ID: errorGroup.ID}}).Updates(&model.ErrorGroup{ImpactfulScore: &totalScore, ImpactfulScoreDate: &currentTime}).Error; err != nil {
+			log.WithContext(ctx).WithField("errorGroupId", errorGroup.ID).Error(err)
+			continue
+		}
+
+		log.WithContext(ctx).WithFields(log.Fields{"errorGroupId": errorGroup.ID, "occurranceScore": occurranceScore, "affectedUsersScore": affectedUsersScore}).Info("Scored error group")
+	}
+}
+
+func (errorScorer *ErrorScorer) scoreErrorGroup(ctx context.Context, errorGroup *ErrorGroupWithLastObjectId) (float64, float64, error) {
+	newLimitedDataError := errorGroup.CreatedAt.After(time.Now().Add(-time.Hour * 24 * 7))
+
+	var occuranceScore sql.NullFloat64
+	var affectedUsersScore sql.NullFloat64
+
+	if newLimitedDataError {
+		// number of occurances
+		if err := errorScorer.db.Raw(`
+			WITH most_occurring_errors AS (
```

Comment:
for these queries, could you do an `explain analyze` against the prod schema to confirm this will use index scans?

Reply:
This also uses index scans, but definitely should add some spans here to help monitor

```
QUERY PLAN | Limit  (cost=48596.56..48596.56 rows=1 width=96) (actual time=1349.670..1349.677 rows=1 loops=1)
-[ RECORD 2 ]-------------------------
QUERY PLAN |   CTE trailing_vals
-[ RECORD 3 ]-------------------------
QUERY PLAN |     ->  GroupAggregate  (cost=48578.24..48580.50 rows=82 width=76) (actual time=1301.365..1342.164 rows=132 loops=1)
-[ RECORD 4 ]-------------------------
QUERY PLAN |           Group Key: (EXTRACT(day FROM error_objects_1.created_at)), (EXTRACT(hour FROM error_objects_1.created_at)), error_objects_1.error_group_id
-[ RECORD 5 ]-------------------------
QUERY PLAN |           ->  Sort  (cost=48578.24..48578.45 rows=82 width=68) (actual time=1301.356..1317.106 rows=60069 loops=1)
-[ RECORD 6 ]-------------------------
QUERY PLAN |                 Sort Key: (EXTRACT(day FROM error_objects_1.created_at)) DESC, (EXTRACT(hour FROM error_objects_1.created_at)) DESC, error_objects_1.error_group_id
-[ RECORD 7 ]-------------------------
QUERY PLAN |                 Sort Method: quicksort  Memory: 4822kB
-[ RECORD 8 ]-------------------------
QUERY PLAN |                 ->  Nested Loop  (cost=48544.35..48575.63 rows=82 width=68) (actual time=1081.743..1201.730 rows=60069 loops=1)
-[ RECORD 9 ]-------------------------
QUERY PLAN |                       ->  Limit  (cost=48543.77..48543.78 rows=5 width=12) (actual time=1081.696..1081.702 rows=5 loops=1)
-[ RECORD 10 ]-------------------------
QUERY PLAN |                             ->  Sort  (cost=48543.77..48572.51 rows=11494 width=12) (actual time=1081.696..1081.699 rows=5 loops=1)
-[ RECORD 11 ]-------------------------
QUERY PLAN |                                   Sort Key: (count(*)) DESC
-[ RECORD 12 ]-------------------------
QUERY PLAN |                                   Sort Method: top-N heapsort  Memory: 25kB
-[ RECORD 13 ]-------------------------
QUERY PLAN |                                   ->  HashAggregate  (cost=48237.92..48352.86 rows=11494 width=12) (actual time=1081.612..1081.676 rows=145 loops=1)
-[ RECORD 14 ]-------------------------
QUERY PLAN |                                         Group Key: error_objects_2.error_group_id
-[ RECORD 15 ]-------------------------
QUERY PLAN |                                         Batches: 1  Memory Usage: 417kB
-[ RECORD 16 ]-------------------------
QUERY PLAN |                                         ->  Index Scan using idx_error_objects_created_at on error_objects error_objects_2  (cost=0.57..48031.19 rows=41346 width=4) (actual time=1.385..1072.648 rows=65526 loops=1)
-[ RECORD 17 ]-------------------------
QUERY PLAN |                                               Index Cond: (created_at > (now() - '7 days'::interval))
-[ RECORD 18 ]-------------------------
QUERY PLAN |                                               Filter: (project_id = 1)
-[ RECORD 19 ]-------------------------
QUERY PLAN |                                               Rows Removed by Filter: 715962
-[ RECORD 20 ]-------------------------
QUERY PLAN |                       ->  Index Only Scan using error_objects_error_group_id_created_at_id_index on error_objects error_objects_1  (cost=0.57..6.12 rows=16 width=12) (actual time=0.021..16.135 rows=12014 loops=5)
-[ RECORD 21 ]-------------------------
QUERY PLAN |                             Index Cond: ((error_group_id = error_objects_2.error_group_id) AND (created_at > (now() - '7 days'::interval)))
-[ RECORD 22 ]-------------------------
QUERY PLAN |                             Heap Fetches: 57721
-[ RECORD 23 ]-------------------------
QUERY PLAN |   ->  Sort  (cost=16.06..16.27 rows=82 width=96) (actual time=1349.670..1349.672 rows=1 loops=1)
-[ RECORD 24 ]-------------------------
QUERY PLAN |         Sort Key: trailing_vals.day DESC, trailing_vals.hour DESC
-[ RECORD 25 ]-------------------------
QUERY PLAN |         Sort Method: top-N heapsort  Memory: 25kB
-[ RECORD 26 ]-------------------------
QUERY PLAN |         ->  Nested Loop  (cost=12.29..15.65 rows=82 width=96) (actual time=1342.313..1349.637 rows=132 loops=1)
-[ RECORD 27 ]-------------------------
QUERY PLAN |               ->  Nested Loop  (cost=12.29..12.37 rows=1 width=72) (actual time=1342.308..1342.313 rows=1 loops=1)
-[ RECORD 28 ]-------------------------
QUERY PLAN |                     ->  Nested Loop  (cost=3.70..3.75 rows=1 width=64) (actual time=1342.285..1342.288 rows=1 loops=1)
-[ RECORD 29 ]-------------------------
QUERY PLAN |                           ->  Aggregate  (cost=1.85..1.86 rows=1 width=32) (actual time=1342.257..1342.258 rows=1 loops=1)
-[ RECORD 30 ]-------------------------
QUERY PLAN |                                 ->  CTE Scan on trailing_vals trailing_vals_1  (cost=0.00..1.64 rows=82 width=8) (actual time=1301.366..1342.202 rows=132 loops=1)
-[ RECORD 31 ]-------------------------
QUERY PLAN |                           ->  Aggregate  (cost=1.85..1.86 rows=1 width=32) (actual time=0.025..0.026 rows=1 loops=1)
-[ RECORD 32 ]-------------------------
QUERY PLAN |                                 ->  CTE Scan on trailing_vals trailing_vals_2  (cost=0.00..1.64 rows=82 width=8) (actual time=0.000..0.011 rows=132 loops=1)
-[ RECORD 33 ]-------------------------
QUERY PLAN |                     ->  Aggregate  (cost=8.60..8.61 rows=1 width=8) (actual time=0.020..0.021 rows=1 loops=1)
-[ RECORD 34 ]-------------------------
QUERY PLAN |                           ->  Index Only Scan using error_objects_error_group_id_created_at_id_index on error_objects  (cost=0.57..8.59 rows=1 width=0) (actual time=0.019..0.019 rows=0 loops=1)
-[ RECORD 35 ]-------------------------
QUERY PLAN |                                 Index Cond: ((error_group_id = 210345215) AND (created_at > (now() - '01:00:00'::interval)))
-[ RECORD 36 ]-------------------------
QUERY PLAN |                                 Heap Fetches: 0
-[ RECORD 37 ]-------------------------
QUERY PLAN |               ->  CTE Scan on trailing_vals  (cost=0.00..1.64 rows=82 width=64) (actual time=0.000..0.013 rows=132 loops=1)
-[ RECORD 38 ]-------------------------
QUERY PLAN | Planning Time: 0.533 ms
-[ RECORD 39 ]-------------------------
QUERY PLAN | Execution Time: 1350.725 ms
EXPLAIN
```